### PR TITLE
Add warnings for registration starting too early

### PR DESCRIPTION
**Pull request by DanielEgdal**
It has happened a few times where a competition was announced and registration opened in less than the required 48 hours.

### CODE REVIEW

File: WcaOnRails/app/models/competition.rb [hunk 1]
```diff
@@ -357,6 +357,10 @@ def warnings_for(user)
         warnings[:name] = I18n.t('competitions.messages.name_too_long')
       end
 
+      if (self.registration_open - Time.now) > 172_800
```
Comment (jonatanklosko) [inline WcaOnRails/app/models/competition.rb:360]: Would make sense to move this value to a constant like [this](https://github.com/thewca/worldcubeassociation.org/blob/16dca929c434e83f15a4d3b7e92cfbca73b026b1/WcaOnRails/app/models/competition.rb#L212).

File: WcaOnRails/app/models/competition.rb [hunk 2]
```diff
@@ -376,6 +380,10 @@ def warnings_for(user)
         warnings[:announcement] = I18n.t('competitions.messages.not_announced')
       end
 
+      if (self.registration_open - self.announced_at) > 172_800
```
Comment (jonatanklosko) [inline WcaOnRails/app/models/competition.rb:383]: If the competition is not announced you subtract `nil` here. I think it would be more clear move this outside the main `if` into an outer one like this:

```rb
if self.announced?
  # one check
else
  # the other check
end
```

**Review by jonatanklosko — COMMENTED** [review]
Some comments from me :>

File: WcaOnRails/app/models/competition.rb (hunk 2 above)
Comment (DanielEgdal) [reply WcaOnRails/app/models/competition.rb:383]: I don't think this issue would error come up. showAtAll is when WCAT has approved the competition and made it visible. The announcement button is pressed a few seconds after.

Note to self: There needs to also be a check for if the results have been posted. There a lot of past competitions which had registration open in less than the currently required 48h after announcement and they don't need to see this error.

I don't have an opinion on your concrete proposal, I just built on top of what was already in the code.

File: WcaOnRails/app/models/competition.rb (hunk 1 above)
Comment (DanielEgdal) [reply WcaOnRails/app/models/competition.rb:360]: It probably would, yes.

File: WcaOnRails/app/models/competition.rb (hunk 2 above)
Comment (jonatanklosko) [reply WcaOnRails/app/models/competition.rb:383]: Between approving a competition and pressing he announcement button there is a refresh, in which case I can imagine the page crashing ^^ Also now I see some tests fail because `registration_open` may be `nil`, so it would be best to check for that. As this is itself a bit of logic, I'd still suggest moving it outside the main `if`.

```rb
# the first condition ensures the registration range fields are other than nil,
# the second condition definitely helps to ignore past competitions, but makes more sense

if registration_range_specified? && !registration_past?
  if self.announced?
    # one check
  else
    # the other check
  end
end

# somewhere below

def registration_range_specified?
  registration_open.present? && registration_close.present?
end
```

Comment (DanielEgdal) [reply WcaOnRails/app/models/competition.rb]: I tried implementing what you sent now! Sorry for the delay, I hope it works this time. :)

File: WcaOnRails/app/models/competition.rb [hunk 3]
```diff
@@ -384,6 +387,22 @@ def warnings_for(user)
     warnings
   end
 
+  def reg_warnings(user)
```
Comment (jonatanklosko) [inline WcaOnRails/app/models/competition.rb:390]: It's great you extracted that to a separate method, although it's not used anywhere now ^^ Also the method doesn't use the `user` argument, so you can remove it. Then it could be used [here](https://github.com/thewca/worldcubeassociation.org/blob/18d3bde220cdf8c804c37c61af2d92975fd40ae5/WcaOnRails/app/models/competition.rb#L386) like this:


```rb
if reg_warnings.any?
  warnings = reg_warnings.merge(warnings)
end
```

File: WcaOnRails/app/models/competition.rb [hunk 4]
```diff
@@ -384,6 +387,22 @@ def warnings_for(user)
...
+        end
+      else
+        if (self.registration_open - Time.now) < REGISTRATION_OPENING_EARLIEST
+          warnings[:regearly] = I18n.t('competitions.messages.reg_opens_too_early')
+        end
+      end
+
+      warnings
+    end
+
```
Comment (jonatanklosko) [inline WcaOnRails/app/models/competition.rb:405]: I believe you're missing an `end` here.

File: WcaOnRails/app/models/competition.rb [hunk 5]
```diff
@@ -950,8 +973,9 @@ def has_location?
     latitude.present? && longitude.present?
   end
 
+  # The division is to convert the end result from secods to days. .to_date removed some hours from the subtraction
   def days_until
-    start_date ? (start_date - Time.now.utc.to_date).to_i : nil
+    start_date ? ((start_date - Time.now.utc)/(86_400)).to_i : nil
```
Comment (jonatanklosko) [inline WcaOnRails/app/models/competition.rb:978]: To be perfectly precise here I believe we should do `start_date.to_time(:utc)`, otherwise `start_date` may be cast to a local `Time` 00:00, which is not what we want here.

File: WcaOnRails/app/models/competition.rb (hunk 3 above)
Comment (DanielEgdal) [reply WcaOnRails/app/models/competition.rb]: I don't really see where it is still being called, but I tried implementing this. It seems this line is now what is causing a few issues for travis.

File: WcaOnRails/app/models/competition.rb [hunk 6]
```diff
@@ -380,10 +383,30 @@ def warnings_for(user)
...
+        if (self.registration_open - self.announced_at) < REGISTRATION_OPENING_EARLIEST
+          warnings[:regearly] = I18n.t('competitions.messages.reg_opens_too_early')
+        end
+      else
+        if (self.registration_open - Time.now.utc) < REGISTRATION_OPENING_EARLIEST
+          warnings[:regearly] = I18n.t('competitions.messages.reg_opens_too_early')
+        end
+      end
+
+      warnings
```
Comment (jonatanklosko) [inline WcaOnRails/app/models/competition.rb:406]: Ah, I've just noticed this is still in the if block, but should be outside, at the end of this method (so we always return a map).

Comment (DanielEgdal) [reply WcaOnRails/app/models/competition.rb]: Nice, I corrected it accordingly and it's no longer failing! :D

**Review by jonatanklosko — APPROVED** [review]
LGTM, thanks!
